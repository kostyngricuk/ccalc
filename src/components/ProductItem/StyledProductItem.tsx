import styled from 'styled-components';
import Button from '@components/UI/Button'

export const StyledProductItem = styled.li`
  margin-bottom: 10px;
  background-color: var(--color-lightgray);
  display: flex;
  align-items: flex-end;
  justify-content: space-between;
  position: relative;
  padding: 30px 40px;
`;

export const StyledRemoveProductButton = styled(Button)`
  color: var(--color-red);
  background-color: transparent;
  text-decoration: underline;
  font-size: 16px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  border: none;
  position: absolute;
  right: 15px;
  top: 10px;
  &:hover,
  &:focus {
    color: var(--color-red);
  }
`;

export default StyledProductItem;
