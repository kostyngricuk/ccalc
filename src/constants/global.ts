export enum LANGUAGES {
  en = 'English',
  ru = 'Russian',
}

export const LANGUAGES_CODE_LIST = Object.keys(LANGUAGES) as Array<keyof typeof LANGUAGES>;

export enum UNITS {
  kg = 'kg',
  g = 'g',
  ml = 'ml',
  sm = 'sm',
}

export const reqSaveCalcAction = 'product/REQUEST_SAVE_CALC';
