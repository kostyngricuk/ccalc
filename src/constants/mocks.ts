import { IProduct, ISelectedProduct, KEY_PRODUCT_ID } from "types/products"
import { Genders, TUser } from "types/user"

export const defaultMockUser: TUser = {
  id: 'unicStringId',
  height: 180,
  weight: 85,
  weightGoal: 80,
  age: 26,
  gender: Genders.man,
  email: 'mockUser@gmail.com',
  calorieWidget: {
    limit: 2300,
    eaten: 1980,
  },
}

export const defaultMockProduct: IProduct = {
  [KEY_PRODUCT_ID]: 0,
  name: 'Milk',
  kkal: 34,
  proto: 1,
  carbo: 3,
  fats: 2,
}

export const defaultMockSelectedProduct: ISelectedProduct = {
  [KEY_PRODUCT_ID]: 0,
  name: 'Milk',
  kkal: 34,
  proto: 1,
  carbo: 3,
  fats: 2,
  weight: 100,
}
