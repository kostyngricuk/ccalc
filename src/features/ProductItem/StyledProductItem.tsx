import styled from 'styled-components';
import Button from 'components/Button'

export const StyledProductItem = styled.li`
  margin-bottom: 10px;
  background-color: var(--color-disabledLight);
  display: flex;
  align-items: flex-end;
  justify-content: space-between;
  position: relative;
  padding: 30px 40px;
`;

export const StyledRemoveProductButton = styled(Button)`
  color: var(--color-third);
  background-color: transparent;
  text-decoration: underline;
  font-size: 16px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  border: none;
  position: absolute;
  right: 15px;
  top: 10px;
  &:hover,
  &:focus {
    color: var(--color-third);
  }
`;

export default StyledProductItem;
