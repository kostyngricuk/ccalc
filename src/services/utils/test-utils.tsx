/* eslint-disable import/no-extraneous-dependencies */
import React, {ReactNode} from 'react'
import lightTheme from '@services/styled/themes'
import { ThemeProvider } from 'styled-components'
import { Provider } from 'react-redux'
import { RootState, setupStore } from '@services/store';
import { Genders, TUser } from '@services/types/user';
import { BrowserRouter } from 'react-router-dom';

export function Providers({
  children,
  state,
}: {
  children: ReactNode,
  state: Partial<RootState>,
}) {
  const store = setupStore(state);
  return (
    <ThemeProvider theme={lightTheme}>
      <Provider store={store}>
        <BrowserRouter>
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>
  )
}

export const defaultMockUser: TUser = {
  id: 'unicStringId',
  height: 180,
  weight: 85,
  weightGoal: 80,
  age: 26,
  gender: Genders.man,
  email: 'mockUser@gmail.com',
  calorieWidget: {
    limit: 2300,
    eaten: 1980,
  },
}
