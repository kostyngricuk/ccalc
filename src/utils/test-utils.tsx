/* eslint-disable import/no-extraneous-dependencies */
import React, {PropsWithChildren, ReactNode} from 'react'
import lightTheme from 'styles/themes'
import { ThemeProvider } from 'styled-components'
import { Provider } from 'react-redux'
import { RootState } from 'store/store';
import { BrowserRouter } from 'react-router-dom';
import { render } from '@testing-library/react'
import type { RenderOptions } from '@testing-library/react'
import configureMockStore, { MockStore, MockStoreCreator } from 'redux-mock-store';
import { I18nextProvider } from 'react-i18next';
import i18n from 'locales/i18nForTests';
import { Action, runSaga, stdChannel } from 'redux-saga';

export interface IDispatchRes {
  type: Action,
  error?: Error
}

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>
}

export function Providers({
  children,
  store,
}: {
  children: ReactNode,
  store: MockStore,
}) {
  return (
    <ThemeProvider theme={lightTheme}>
      <Provider store={store}>
        <I18nextProvider i18n={i18n} defaultNS="translation">
          <BrowserRouter>
            {children}
          </BrowserRouter>
        </I18nextProvider>
      </Provider>
    </ThemeProvider>
  )
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  const mockStore: MockStoreCreator = configureMockStore();
  const store: MockStore = mockStore(preloadedState);
  return {
    store,
    ...render(
      ui,
      {
        wrapper: ({ children }: PropsWithChildren<{}>) => (
          <Providers store={store}>
            { children}
          </Providers>
        ),
        ...renderOptions
      }
    )
  }
}

export const runMockSaga = async (sagaFunc: any, sagaData: object, dispatchRes?: any) => {
  const channel = stdChannel()
  await runSaga(
    {
      channel,
      dispatch: (res) => {
        if (dispatchRes) {
          dispatchRes.push(res)
        }
      },
    },
    () => sagaFunc(sagaData));
}

export const createMockApiRequest = (api: any, apiMethod: any, mockResponse: object) => jest.spyOn(api, apiMethod)
  .mockImplementation(() => Promise.resolve(mockResponse))

// re-export everything
export * from '@testing-library/react'

// override render method
export {renderWithProviders as render}