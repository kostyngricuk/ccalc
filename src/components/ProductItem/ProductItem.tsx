import React, { useEffect, useState } from "react";
import { useTranslation } from "react-i18next";

import { IProduct } from "../../services/types/products";
import { StyledProductItem, StyledRemoveProductButton } from "./StyledProductItem";
import Title from "../UI/Title/Title";
import { EnumInputType, Input } from "../UI/Input/Input";
import { UNITS } from "../../services/constants/global";
import { EnumTitleVariant } from "../../services/types/global";
import { useAppDispatch } from "../../services/hooks/store";
import { removeProduct } from "../../services/reducers/productSlice";
import { CloseSVG } from "../../icons";
import Icon from "../UI/Icon/Icon";

export default function ProductItem({
  item
}: {
  item: IProduct
}) {
  const [weigh, setWeight] = useState<number>(100);
  const [carbo, setCarbo] = useState<number>(item.carbo);
  const [fats, setFats] = useState<number>(item.fats);
  const [proto, setProto] = useState<number>(item.proto);
  const [kkal, setKkal] = useState<number>(item.kkal);
  const { t } = useTranslation();
  const dispatch = useAppDispatch();


  useEffect(() => {
    if (!weigh) {
      return;
    }
    setCarbo(parseFloat((item.carbo / 100 * weigh).toFixed(2)));
    setFats(parseFloat((item.fats / 100 * weigh).toFixed(2)));
    setProto(parseFloat((item.proto / 100 * weigh).toFixed(2)));
    setKkal(parseFloat((item.kkal / 100 * weigh).toFixed(2)));
  }, [weigh]);

  const removeProductFromList = (itemId: number) => {
    dispatch({
      type: removeProduct.type,
      payload: {
        id: itemId
      }
    });
  }

  return (
    <StyledProductItem>
      <div>
        <Title variant={EnumTitleVariant.h3}>{item.name}</Title>
        <p><strong>{t('calculator.item.nutrients')} :</strong> {proto}/{fats}/{carbo}</p>
        <p><strong>{t('calculator.item.colories')} :</strong> {kkal}</p>
      </div>
      <Input
        units={t(`units.${UNITS.g}`)}
        name="weight"
        value={weigh.toString()}
        label={t('form.field.weight')}
        type={EnumInputType.number}
        onChange={(value: string) => setWeight(parseInt(value, 10))}
      />
      <StyledRemoveProductButton
        $isIcon
        $isOutline
        ariaLabel={t('calculator.item.remove')}
        onClick={() => removeProductFromList(item.id)}
      >
        <Icon Sprite={CloseSVG} />
      </StyledRemoveProductButton>
    </StyledProductItem>
  )
}
