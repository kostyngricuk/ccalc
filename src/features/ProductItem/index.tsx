import React from 'react';
import { useTranslation } from 'react-i18next';

import { ISelectedProduct } from 'types/products';
import { UNITS } from 'constants/global';
import { useAppDispatch } from 'hooks/store';
import { removeProduct, updateProductWeight } from 'store/slices/productSlice';
import { CloseSVG } from 'icons/index';
import { Input } from 'components/Input'
import Title from 'components/Title'
import Icon from 'components/Icon'
import { StyledProductItem, StyledRemoveProductButton } from './StyledProductItem';

export default function ProductItem({
  item
}: {
  item: ISelectedProduct
}) {
  const { t } = useTranslation();
  const dispatch = useAppDispatch();


  const removeProductFromList = (itemId: number) => {
    dispatch({
      type: removeProduct.type,
      payload: {
        id: itemId
      }
    });
  }

  const changeWeight = (value: string) => {
    dispatch({
      type: updateProductWeight.type,
      payload: {
        id: item.id,
        newWeight: parseInt(value, 10)
      }
    });
  }

  const {
    id,
    name,
    kkal,
    proto,
    fats,
    carbo,
    weight
  } = item;

  return (
    <StyledProductItem>
      <div>
        <Title variant='h3'>{name}</Title>
        <p><strong>{t('calculator.item.nutrients')} :</strong> {proto}/{fats}/{carbo}</p>
        <p><strong>{t('calculator.item.colories')} :</strong> {kkal}</p>
      </div>
      <Input
        units={t(`units.${UNITS.g}`)}
        name="weight"
        value={weight.toString()}
        label={t('form.field.weight')}
        type='number'
        onChange={(value: string) => changeWeight(value)}
      />
      <StyledRemoveProductButton
        $isIcon
        $isOutline
        ariaLabel={t('calculator.item.remove')}
        onClick={() => removeProductFromList(id)}
      >
        <Icon sprite={CloseSVG} />
      </StyledRemoveProductButton>
    </StyledProductItem>
  )
}
