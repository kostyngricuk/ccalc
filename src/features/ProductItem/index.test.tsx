import React from "react";
import { render } from "utils/test-utils";
import ProductItem from "features/ProductItem";
import { defaultMockSelectedProduct } from "constants/mocks";

describe('Product Item', () => {
  it('Should render with default props', () => {
    const wrapper = render(<ProductItem item={defaultMockSelectedProduct} />);

    expect(wrapper.getByRole('heading')).toHaveTextContent(defaultMockSelectedProduct.name);
  })

  it('Should make a snapshot', () => {
    const wrapper = render(<ProductItem item={defaultMockSelectedProduct} />);
    expect(wrapper).toMatchSnapshot();
  })

  it('Should have a remove product button', async () => {
    const wrapper = render(<ProductItem item={defaultMockSelectedProduct} />);

    const removeBtn = wrapper.getByRole('button');
    expect(removeBtn).toBeInTheDocument();
  })

  it('Should have input for changing waight', async () => {
    const wrapper = render(<ProductItem item={defaultMockSelectedProduct} />);

    const weightInput = wrapper.getByRole('spinbutton');
    expect(weightInput).toBeInTheDocument();
  })
})