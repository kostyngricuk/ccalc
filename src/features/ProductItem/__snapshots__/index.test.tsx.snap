// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Item Should make a snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="sc-gsFSjX bxkOIO"
      >
        <div>
          <h3
            class="sc-fqkwJk kpHTiC"
            position="left"
          >
            Milk
          </h3>
          <p>
            <strong>
              P / F / C
               :
            </strong>
             
            1
            /
            2
            /
            3
          </p>
          <p>
            <strong>
              Сalories
               :
            </strong>
             
            34
          </p>
        </div>
        <div
          class="sc-aYaIB eLKLGl Input"
        >
          <label
            class="sc-gEvDqW cDSHgZ has-units"
          >
            <input
              autocomplete="off"
              data-testid="input"
              name="weight"
              type="number"
              value="100"
            />
            <span
              class="label"
            >
              Weight
            </span>
            <span
              class="units"
            >
              g
            </span>
          </label>
        </div>
        <button
          aria-label="Remove product"
          class="sc-iGgVNO bmyEPv sc-kAycRU ilFrKt"
          color="primary"
          type="button"
        >
          <span
            class="sc-dcJtft gzkRHp Icon"
            height="24px"
            role="img"
            width="24px"
          >
            <div>
              <div>
                <div
                  data-src="div"
                  src=""
                />
              </div>
            </div>
          </span>
        </button>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="sc-gsFSjX bxkOIO"
    >
      <div>
        <h3
          class="sc-fqkwJk kpHTiC"
          position="left"
        >
          Milk
        </h3>
        <p>
          <strong>
            P / F / C
             :
          </strong>
           
          1
          /
          2
          /
          3
        </p>
        <p>
          <strong>
            Сalories
             :
          </strong>
           
          34
        </p>
      </div>
      <div
        class="sc-aYaIB eLKLGl Input"
      >
        <label
          class="sc-gEvDqW cDSHgZ has-units"
        >
          <input
            autocomplete="off"
            data-testid="input"
            name="weight"
            type="number"
            value="100"
          />
          <span
            class="label"
          >
            Weight
          </span>
          <span
            class="units"
          >
            g
          </span>
        </label>
      </div>
      <button
        aria-label="Remove product"
        class="sc-iGgVNO bmyEPv sc-kAycRU ilFrKt"
        color="primary"
        type="button"
      >
        <span
          class="sc-dcJtft gzkRHp Icon"
          height="24px"
          role="img"
          width="24px"
        >
          <div>
            <div>
              <div
                data-src="div"
                src=""
              />
            </div>
          </div>
        </span>
      </button>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "clearActions": [Function],
    "dispatch": [Function],
    "getActions": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
